CONTRIBUTING.md

```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
+
+Each log message is written at a particular _level_. By default Elasticsearch
+will suppress messages at the two most verbose levels, `TRACE` and `DEBUG`, and
+will output messages at all other levels. Users can configure which levels of
+message are written by each logger at runtime, but you should expect everyone
+to run with the default configuration almost all of the time and choose your
+levels accordingly.
+
+##### `TRACE`
+
+This is the most verbose level, disabled by default, and it is acceptable if it
+generates a very high volume of logs. The target audience of `TRACE` logs
+comprises developers who are trying to deeply understand some unusual runtime
+behaviour of a system. For instance `TRACE` logs may be useful when
+understanding an unexpected interleaving of concurrent actions or some
+unexpected consequences of a delayed response from a remote node.
+
+`TRACE` logs will normally only make sense when read alongside the code, and
+typically they will be read as a whole sequence of messages rather than in
+isolation.
+
+Even though `TRACE` logs may be very verbose, you should still exercise some
+judgement when deciding when to use them. In many cases it will be easier to
+understand the behaviour of the system using tests or by analysing the code
+itself rather than by trawling through hundreds of trivial log messages.
+
+It may not be easy, or even possible, to obtain `TRACE` logs from a production
+system. Therefore they are not appropriate for information that you would
+expect to be useful in diagnosing problems in production.
+
+##### `DEBUG`
+
+This is the next least verbose level and is also disabled by default. The
+target audience of this level typically comprises users or developers who are
+trying to diagnose an unexpected problem in a production system, perhaps to
+help determine whether a fault lies within Elasticsearch or elsewhere.
+
+Users should expect to be able to enable `DEBUG` logging on their production
+systems for a whole subsystem for an extended period of time without
+overwhelming the system or filling up their disks with logs, so it is important
+to limit the volume of messages logged at this level. On the other hand, these
+messages must still provide enough detail to diagnose the sorts of problems
+that Elasticsearch might encounter. In some cases it works well to collect
+information over a period of time and then log a complete summary, rather than
+recording every step of a process in its own message.
+
+It's possible that the reader of `DEBUG` logs is also reading the code, but
+that is less likely than for `TRACE` logs. You should strive to avoid
+terminology that only makes sense when reading the code, and also aim for
+messages at this level to be self-contained rather than intending them to be
+read as a sequence.
+
+It's often useful to log exceptions and other deviations from the "happy path"
+at `DEBUG` level. Exceptions logged at `DEBUG` should generally include the
+complete stack trace.
```

Comment:
It might be useful to call out a few things that shouldn't be logged at DEBUG level.

Reply:
We already mention thinking about (a) volume and (b) self-contained messages vs sequences. What sort of thing were you thinking of here?

Reply:
I was mostly thinking about concrete examples. Maybe with the whole structure of the document we could add some examples at the end and explain the reasoning. Sort of giving an example of applying the rules you've laid out. Its fine not to have it though.

Reply:
++ I added some examples in 5cf06e4f1c3560edfee7c4f6e5d443fbb27b2d15.